pkg/alertmanager/multitenant_test.go

```diff
@@ -601,6 +608,97 @@ func TestMultitenantAlertmanager_deleteUnusedLocalUserState(t *testing.T) {
+		reg := prometheus.NewPedanticRegistry()
+		cfg := mockAlertmanagerConfig(t)
+
+		cfg.ShardingRing.ReplicationFactor = 2
+		cfg.ShardingRing.InstanceID = fmt.Sprintf("instance-%d", i)
+		cfg.ShardingRing.InstanceAddr = fmt.Sprintf("127.0.0.1-%d", i)
+		cfg.ShardingEnabled = true
+		cfg.ShardingRing.ZoneAwarenessEnabled = true
+		cfg.ShardingRing.InstanceZone = zone
+
+		// Increase state write interval so that state gets written sooner, making test faster.
+		cfg.Persister.Interval = 500 * time.Millisecond
+
+		am, err := createMultitenantAlertmanager(cfg, nil, nil, alertStore, ringStore, nil, log.NewLogfmtLogger(os.Stdout), reg)
+		require.NoError(t, err)
+		t.Cleanup(func() {
+			require.NoError(t, services.StopAndAwaitTerminated(ctx, am))
+		})
+		require.NoError(t, services.StartAndAwaitRunning(ctx, am))
+
+		return am
+	}
+
+	am1ZoneA := createInstance(1, "zoneA")
+	am2ZoneA := createInstance(2, "zoneA")
+	am1ZoneB := createInstance(3, "zoneB")
+
+	// Configure the users and wait for the state persister to write some state for both.
+	{
+		require.NoError(t, alertStore.SetAlertConfig(ctx, alertspb.AlertConfigDesc{
+			User:      user1,
+			RawConfig: simpleConfigOne,
+			Templates: []*alertspb.TemplateDesc{},
+		}))
+		require.NoError(t, alertStore.SetAlertConfig(ctx, alertspb.AlertConfigDesc{
+			User:      user2,
+			RawConfig: simpleConfigOne,
+			Templates: []*alertspb.TemplateDesc{},
+		}))
+		require.NoError(t, alertStore.SetAlertConfig(ctx, alertspb.AlertConfigDesc{
+			User:      user3,
+			RawConfig: simpleConfigOne,
+			Templates: []*alertspb.TemplateDesc{},
+		}))
+
+		err := am1ZoneA.loadAndSyncConfigs(context.Background(), reasonPeriodic)
+		require.NoError(t, err)
+		err = am2ZoneA.loadAndSyncConfigs(context.Background(), reasonPeriodic)
+		require.NoError(t, err)
+		err = am1ZoneB.loadAndSyncConfigs(context.Background(), reasonPeriodic)
+		require.NoError(t, err)
+
+		require.Eventually(t, func() bool {
+			_, err1 := alertStore.GetFullState(context.Background(), user1)
+			_, err2 := alertStore.GetFullState(context.Background(), user2)
+			_, err3 := alertStore.GetFullState(context.Background(), user3)
+			return err1 == nil && err2 == nil && err3 == nil
+		}, 5*time.Second, 100*time.Millisecond, "timed out waiting for state to be persisted")
+	}
+
+	zoneATotal := 0
+	allUserIDs, configs, err := am1ZoneA.loadAlertmanagerConfigs(ctx)
+	require.NoError(t, err)
+	assert.Equal(t, len(allUserIDs), 3)
+	zoneATotal += len(configs)
+
+	allUserIDs, configs, err = am2ZoneA.loadAlertmanagerConfigs(ctx)
+	require.NoError(t, err)
+	assert.Equal(t, len(allUserIDs), 3)
+	zoneATotal += len(configs)
+
+	assert.Equal(t, len(allUserIDs), zoneATotal)
+
+	allUserIDs, configs, err = am1ZoneB.loadAlertmanagerConfigs(ctx)
+	require.NoError(t, err)
+	assert.Equal(t, len(allUserIDs), 3)
+	// This is tis only instance on this AZ, so all users should be loaded on this instance
+	assert.Equal(t, len(configs), len(allUserIDs))
```

Comment:
For these assertions, I see the logic of using the `loadAlertmanagerConfigs` as a way of checking what users are owned by the instance, but relying on the internal function I think should be a last resort, it will make it harder to e.g. refactor in the future.

In this case using the metrics to test would be nice as a) it's user facing and b) we're also testing the metrics.

```
	registriesZoneA := util.NewUserRegistries()
	registriesZoneB := util.NewUserRegistries()

	// ... add reg for instance 1,2 to A / 3 to B with AddUserRegistry

	assert.Equal(t, float64(3), regZoneA.GetSumOfGauges("cortex_alertmanager_tenants_owned"))
	assert.Equal(t, float64(3), regZoneB.GetSumOfGauges("cortex_alertmanager_tenants_owned"))	
```

Reply:
This ☝️  way of testing metric is fine, but using `testutil.GatherAndCompare()` is even easier and cleaner (see other Cortex tests asserting on `testutil.GatherAndCompare()`). You just have to return the registry from `createInstance()` so you can assert on the right one.

Reply:
Oh yeah, you're right!